In [1]:
import pandas as pd;
import numpy as np;

In [3]:
# load data

sf_permits = pd.read_csv("/home/nbuser/library/Building_Permits.csv");


/home/nbuser/anaconda3_501/lib/python3.6/site-packages/IPython/core/interactiveshell.py:2728: DtypeWarning: Columns (22,32) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [8]:
sf_permits.head()

#head gives you specified number of records from top. Degfault value is 5 and can be overriiden by passing parameter to head.

Permit Number  Permit Type            Permit Type Definition  \
0  201505065519            4                      sign - erect   
1  201604195146            4                      sign - erect   
2  201605278609            3  additions alterations or repairs   
3  201611072166            8            otc alterations permit   
4  201611283529            6                       demolitions   

  Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
0           05/06/2015  0326  023            140                  NaN   
1           04/19/2016  0306  007            440                  NaN   
2           05/27/2016  0595  203           1647                  NaN   
3           11/07/2016  0156  011           1230                  NaN   
4           11/28/2016  0342  001            950                  NaN   

  Street Name Street Suffix      ...        Existing Construction Type  \
0       Ellis            St      ...                               3.0   
1       Geary            St      ...                               3.0   
2     Pacific            Av      ...                               1.0   
3     Pacific            Av      ...                               5.0   
4      Market            St      ...                               3.0   

  Existing Construction Type Description Proposed Construction Type  \
0                          constr type 3                        NaN   
1                          constr type 3                        NaN   
2                          constr type 1                        1.0   
3                         wood frame (5)                        5.0   
4                          constr type 3                        NaN   

  Proposed Construction Type Description Site Permit Supervisor District  \
0                                    NaN         NaN                 3.0   
1                                    NaN         NaN                 3.0   
2                          constr type 1         NaN                 3.0   
3                         wood frame (5)         NaN                 3.0   
4                                    NaN         NaN                 6.0   

  Neighborhoods - Analysis Boundaries  Zipcode  \
0                          Tenderloin  94102.0   
1                          Tenderloin  94102.0   
2                        Russian Hill  94109.0   
3                            Nob Hill  94109.0   
4                          Tenderloin  94102.0   

                                    Location      Record ID  
0  (37.785719256680785, -122.40852313194863)  1380611233945  
1   (37.78733980600732, -122.41063199757738)  1420164406718  
2    (37.7946573324287, -122.42232562979227)  1424856504716  
3   (37.79595867909168, -122.41557405519474)  1443574295566  
4   (37.78315261897309, -122.40950883997789)   144548169992  

[5 rows x 43 columns]

In [7]:
# shape gives you number of rows and columns
sf_permits.shape

#sf_permits has whoppping 198900 rows and 43 columns... imagine analyzing this data manually !

(198900, 43)

In [10]:
sf_permits.sample(10)

#unlike head, sample randomly selects specified data rows from dataset. Default is 1.

Permit Number  Permit Type  Permit Type Definition  \
12384   201305136783            8  otc alterations permit   
147490       M754367            8  otc alterations permit   
149800  201702149387            6             demolitions   
31480        M445067            8  otc alterations permit   
97310        M619008            8  otc alterations permit   
18362   201307111655            8  otc alterations permit   
33893        M451307            8  otc alterations permit   
94914   201508103820            8  otc alterations permit   
154458  201704123857            8  otc alterations permit   
103901  201510281042            8  otc alterations permit   

       Permit Creation Date Block  Lot  Street Number Street Number Suffix  \
12384            05/13/2013  1007  033           3240                  NaN   
147490           01/17/2017  4884  025           1120                  NaN   
149800           02/14/2017  3754  029             40                  NaN   
31480            11/18/2013  1384  019           2331                  NaN   
97310            09/01/2015  4103  051            507                  NaN   
18362            07/11/2013  2632  011           1467                  NaN   
33893            12/13/2013  3732  140            935                  NaN   
94914            08/10/2015  4270  019           3088                  NaN   
154458           04/12/2017  1231  049           1619                  NaN   
103901           10/28/2015  0154  011           1426                  NaN   

        Street Name Street Suffix      ...        Existing Construction Type  \
12384    Sacramento            St      ...                               5.0   
147490       Egbert            Av      ...                               NaN   
149800    Cleveland            St      ...                               5.0   
31480          Lake            St      ...                               NaN   
97310   Mississippi            St      ...                               NaN   
18362       Willard            St      ...                               5.0   
33893        Howard            St      ...                               NaN   
94914          25th            St      ...                               5.0   
154458         Page            St      ...                               5.0   
103901      Pacific            Av      ...                               5.0   

       Existing Construction Type Description Proposed Construction Type  \
12384                          wood frame (5)                        5.0   
147490                                    NaN                        NaN   
149800                         wood frame (5)                        NaN   
31480                                     NaN                        NaN   
97310                                     NaN                        NaN   
18362                          wood frame (5)                        5.0   
33893                                     NaN                        NaN   
94914                          wood frame (5)                        5.0   
154458                         wood frame (5)                        5.0   
103901                         wood frame (5)                        5.0   

       Proposed Construction Type Description Site Permit Supervisor District  \
12384                          wood frame (5)         NaN                 2.0   
147490                                    NaN         NaN                10.0   
149800                                    NaN         NaN                 6.0   
31480                                     NaN         NaN                 1.0   
97310                                     NaN         NaN                10.0   
18362                          wood frame (5)         NaN                 5.0   
33893                                     NaN         NaN                 6.0   
94914                          wood frame (5)         NaN                 9.0   
154458                       

In [11]:
# specify the seed for your random numbers

np.random.seed(0);

#this ensures that same random numbers are generated each time so running np.random.rand(30) will generate same random number each time

In [23]:
#missingvalues_count = sf_permits.isnull() simply gives that whether each data cell is null(true) or not (False)

missingvalues_count = sf_permits.isnull().sum() # this gives the count of null by column name


In [24]:
missingvalues_count

#Make a note of nulls some columns like  Street Number Suffix and Voluntary Soft-Story Retrofit  are full of nulls. The next step is to think how we handle them.

Permit Number                                  0
Permit Type                                    0
Permit Type Definition                         0
Permit Creation Date                           0
Block                                          0
Lot                                            0
Street Number                                  0
Street Number Suffix                      196684
Street Name                                    0
Street Suffix                               2768
Unit                                      169421
Unit Suffix                               196939
Description                                  290
Current Status                                 0
Current Status Date                            0
Filed Date                                     0
Issued Date                                14940
Completed Date                            101709
First Construction Document Date           14946
Structural Notification                   191978
Number of Existing S

In [36]:
total_cells = np.product(sf_permits.shape) # shape returns a 1*2 matrix of rows and columns. np.product multiplies them to get total cells
total_cells

8552700

In [37]:
missing_cells= missingvalues_count.sum() # here using shape does not make sense as we want to get total number of null cells. So we add up the values inach column.

missing_cells

2245941

In [39]:
 percent_missing = (missing_cells/total_cells)*100
    
percent_missing # 26% cells are null or nan.

26.260023150584029

In [51]:
# Let's analyze each cell on why its missing. Let's get missing column names
missingvalues_count[missingvalues_count  >0] # 31 columns have null values

Street Number Suffix                      196684
Street Suffix                               2768
Unit                                      169421
Unit Suffix                               196939
Description                                  290
Issued Date                                14940
Completed Date                            101709
First Construction Document Date           14946
Structural Notification                   191978
Number of Existing Stories                 42784
Number of Proposed Stories                 42868
Voluntary Soft-Story Retrofit             198865
Fire Only Permit                          180073
Permit Expiration Date                     51880
Estimated Cost                             38066
Revised Cost                                6066
Existing Use                               41114
Existing Units                             51538
Proposed Use                               42439
Proposed Units                             50911
Plansets            

# Possible reasons why cells are empty
			Street Number Suffix                      196684 Not all streets has uffixes. Possibly value does not exist.
			Street Suffix                               2768 Possible not recorded
			Unit                                      169421 Possibly value does not exist.
			Unit Suffix                               196939 Possibly value does not exist.
			Description                                  290 Not recorded
			Issued Date                                14940 Not recorded
			Completed Date                            101709 Not exist i.e. undergoing construction
			First Construction Document Date           14946 Not recorded
			Structural Notification                   191978 Not exists as may not e applaicable for all buildings.
			Number of Existing Stories                 42784 Not exists as construction may be underway
			Number of Proposed Stories                 42868  Not exist as plans yet to be finalized.
			Voluntary Soft-Story Retrofit             198865  Not recorded. Can default to N.
			Fire Only Permit                          180073 No recorded Can default to N
			Permit Expiration Date                     51880 Not recorded
			Estimated Cost                             38066 Not recorded
			Revised Cost                                6066 Not recorded
			Existing Use                               41114 Not exist
			Existing Units                             51538 Not exist
			Proposed Use                               42439 Not recorded
			Proposed Units                             50911 Not recorded
			Plansets                                   37309 Not recorded
			TIDF Compliance                           198898 Not recorded
			Existing Construction Type                 43366 Not recorded
			Existing Construction Type Description     43366 Not recorded
			Proposed Construction Type                 43162 Not recorded
			Proposed Construction Type Description     43162 Not recorded
			Site Permit                               193541 Not recorded
			Supervisor District                         1717 Not recorded
			Neighborhoods - Analysis Boundaries         1725 Not recorded
			Zipcode                                     1716 Not recorded
			Location                                    1700 Not recorded

In [57]:
#Option1 : Drop all columns with nulls
option1_dropna = sf_permits.dropna(axis =1)  
# this leaves 12 columns which is expected as we had total of 43 columns and 31 had one or more null values. 
# Axis =0 would have meant that drop any rows which have any nulls.

option1_dropna.shape

(198900, 12)

In [63]:
sf_permits.shape

(198900, 43)

In [65]:
#Option2 : Fill all columns with null with 0

option2_fill0 = sf_permits.fillna(0);

# Now check does option2 leave any nulls

nullcount = option2_fill0.isnull().sum();

nullcount

Permit Number                             0
Permit Type                               0
Permit Type Definition                    0
Permit Creation Date                      0
Block                                     0
Lot                                       0
Street Number                             0
Street Number Suffix                      0
Street Name                               0
Street Suffix                             0
Unit                                      0
Unit Suffix                               0
Description                               0
Current Status                            0
Current Status Date                       0
Filed Date                                0
Issued Date                               0
Completed Date                            0
First Construction Document Date          0
Structural Notification                   0
Number of Existing Stories                0
Number of Proposed Stories                0
Voluntary Soft-Story Retrofit   

In [68]:
#Option3: Fill all null values with values in the row before them

option3_fillwithvaluebefore = sf_permits.fillna(method = "bfill", axis=0)

option3_fillwithvaluebefore.isnull().sum()

#at this point you notice that certain values are still left as null. The reason is that there may not be any values before them to fill.
# So add fillna(0) to fill such values as well.

option3_fillwithvaluebefore = sf_permits.fillna(method = "bfill", axis=0).fillna(0)

option3_fillwithvaluebefore.isnull().sum()

Permit Number                             0
Permit Type                               0
Permit Type Definition                    0
Permit Creation Date                      0
Block                                     0
Lot                                       0
Street Number                             0
Street Number Suffix                      0
Street Name                               0
Street Suffix                             0
Unit                                      0
Unit Suffix                               0
Description                               0
Current Status                            0
Current Status Date                       0
Filed Date                                0
Issued Date                               0
Completed Date                            0
First Construction Document Date          0
Structural Notification                   0
Number of Existing Stories                0
Number of Proposed Stories                0
Voluntary Soft-Story Retrofit   

# Todo
Handling Missing Values: In this notebook Dan shows you several approaches to imputing missing data using scikit-learn's imputer.
Look back at the Zipcode column in the sf_permits dataset, which has some missing values. How would you go about figuring out what the actual zipcode of each address should be? (You might try using another dataset. You can search for datasets about San Fransisco on the Datasets listing.)